src/internal/connector/onedrive/collections.go

```diff
@@ -405,13 +406,49 @@ func (c *Collections) UpdateCollections(
 
 		switch {
 		case item.GetFolder() != nil, item.GetPackage() != nil:
+			var (
+				prevPath path.Path
+				col      data.BackupCollection
+				found    bool
+			)
+
+			prevColPath, ok := oldPaths[*item.GetId()]
+			if ok {
+				prevPath, err = path.FromDataLayerPath(prevColPath, false)
+				if err != nil {
+					logger.Ctx(ctx).Errorw("invalid path for deleted item", "error", err)
+					return err
+				}
+			}
+
 			if item.GetDeleted() != nil {
+				if prevPath == nil {
+					// It is possible that an item was created and
+					// deleted between two delta invocations. In
+					// that case, it will only produce a single
+					// delete entry in the delta response.
+					continue
+				}
+
 				// Nested folders also return deleted delta results so we don't have to
 				// worry about doing a prefix search in the map to remove the subtree of
 				// the deleted folder/package.
 				delete(newPaths, *item.GetId())
 
-				// TODO(ashmrtn): Create a collection with state Deleted.
+				col := NewCollection(
+					c.itemClient,
+					nil,
+					prevPath,
+					driveID,
+					c.service,
+					c.statusUpdater,
+					c.source,
+					c.ctrl,
+					invalidPrevDelta,
+				)
+
+				c.CollectionMap[prevColPath] = col
```

Comment:
I think indexing `CollectionMap` with the previous path could cause issues in some cases. I think we may hit a problem if we encounter the following:
1. delete known folder at path `/foo`
2. create new folder at path `/foo`

In that case, the newly created folder will clobber the entry for the deleted one. The folder itself will end up with the correct items in the snapshot (new folders shouldn't be merged) but it will cause any subfolders of the old folder to be added to the snapshot instead of being deleted (e.x. if a folder `/foo/bar` existed it will be a subfolder of the new `/foo` folder instead of being part of the deletion of `/foo`)

Reply:
Hmm, indexing by name is starting to look more and more flaky. Let me switch to indexing by id. The reason why I was reluctant initially was because we were creating collection when we first encounter an item(this is done so as to avoid empty collections), but at that point the id of a collection is not available. We could store somewhere else and use it when we first find an item, but my worry was weather graph would serve us with a item before the corresponding directory (I have not seen this happen yet).

Reply:
Thinking about we, we might have to start ordering the collections in the order of how they appear. I was initially thinking we could add last modified timestamp to collections, but deletes do not have timestamp associated with them. I think out best bet would be to keep a list of ids in the order of when we encounter them and use that to decide the order in which we return the final set of collections.

Reply:
Opened https://github.com/alcionai/corso/pull/2460 for ordering.

Reply:
as we discussed out of band, kopia is order insensitive as it only examines the previous and current paths of a collection. This means it can't do "tree operations" on collections because it assumes collections know their current location. Instead, we'll need to fixup the path for collections that we already have if one of the parent collections moves. For example, if we have `/foo/bar` because we saw an item for it and then we see the equivalent of `mv /foo /baz` we'll need to update the `FullPath()` of `/foo/bar` to be `/baz/bar`

Since we now have IDs for each collection each map lookup should be easier. The new path of the collection after the parent move will probably be in `newPaths` as that's updated below

I don't think a variable to track the paths we've seen or the order collections arrive in will be required